mcorpc/authz_rego.go

```diff
@@ -0,0 +1,131 @@
+package mcorpc
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/choria-io/go-choria/choria"
+	"github.com/choria-io/go-config"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/topdown"
+	"github.com/sirupsen/logrus"
+)
+
+type regoPolicy struct {
+	cfg   *config.Config
+	req   *Request
+	agent *Agent
+	log   *logrus.Entry
+}
+
+func regoPolicyAuthorize(req *Request, agent *Agent, log *logrus.Entry) (bool, error) {
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:   agent.Config,
+		req:   req,
+		agent: agent,
+		log:   logger,
+	}
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	buf := topdown.NewBufferTracer()
+
+	options := []func(*rego.Rego){
+		rego.Query("data.choria.mcorpc.authpolicy.allow"),
+		rego.Module("//choria/mcorpc/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	}
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		options = append(options, rego.Tracer(buf))
+	}
+
+	query := rego.New(
+		options...,
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	if r.log.Logger.GetLevel() == logrus.TraceLevel {
+		topdown.PrettyTrace(r.log.Writer(), *buf)
+	}
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
```

Comment:
should probably return false here too

Reply:
I considered adding this to the aaa server too, but the lines it makes from the basic error print is actually quite useful already

```
     could not evaluate rego policy: 1 error occurred: 1 error occurred: choria.rego:16: rego_parse_error: no match found
            input.claims.callerid !8Z "up=bob"
```

deliberate error at `!8Z`

Reply:
```
      Unexpected error:
          <ast.Errors | len:1, cap:1>: [
              {
                  Code: "rego_unsafe_var_error",
                  Message: "var asdf is unsafe",
                  Location: {
                      Text: "asdf",
                      File: "//choria/mcorpc/policies/facts",
                      Row: 8,
                      Col: 5,
                  },
                  Details: nil,
              },
          ]
          1 error occurred: //choria/mcorpc/policies/facts:8: rego_unsafe_var_error: var asdf is unsafe
      occurred
```